### DISCUSSION TITLE
Working with Fastmail / Overture JavaScript functions

### DISCUSSION THREAD

**Discussion by mdbraber** [joelekstrom/fastmate#89] [Ideas]
I'm trying to add functions to add to Fastmate, e.g. a separate compose window. To achieve this I've looked into ways to 'tie in' to FastMail JavaScript functions (Fastmail has published their frontend JS framework Overture: https://overturejs.com).

Below is an example how to create your own version of what happens when you click the "Discard" button (e.g. in this case it pops up a dialog to confirm).

```
Function.prototype.clone = function() {
    var that = this;
    var temp = function temporary() { return that.apply(this, arguments); };
    for(var key in this) {
        if (this.hasOwnProperty(key)) {
            temp[key] = this[key];
        }
    }
    return temp;
};

let discard_timer = setInterval(function() {
    try {
        let fn = FastMail.classes.ComposeController.prototype.discard.clone();
            FastMail.classes.ComposeController.prototype.discard = function(t) {
            if (!confirm("Press OK to discard email")) {
                return;
            }
            return fn.apply(this, Array.prototype.slice.call(arguments));
        }
        clearInterval(discard_timer);
    } catch (error) {}
}, 500);
```

I'm still looking for ways to 'call' FastMail functions, such as the `undo()` function directly. There is a `Fastmail` object in JS, but I don't know (yet) e.g. how to instantiate classes like the `UndoManager` or `StoreUndoManager`. Any tips or suggestions welcome.

Some pointers to a bit more advanced Fastmail JS functions: https://github.com/rjbs/misc/blob/main/fastamper.js and https://github.com/lamby/fastmail-enhancement-suite/blob/master/js/main.js

**Comment by joelekstrom** [comment]
I'm not a JS dev and my knowledge here is very limited, so I'm not of much help here unfortunately 😞 

Looks like you're on to something there with overriding their functions like `FastMail.classes.ComposeController.prototype.discard`, but in the end the trouble here is that they need to have an official/stable API if this is to keep working as intended, otherwise it has kind of the same problems as the HTML class checking we do currently. JS functions are likely much more stable than HTML classes though so it would be an improvement 👍

**Reply by mdbraber** [reply]
Same here, I mostly piece together from other smart people. Don’t think it is as unstable as the `div` ids and at overturejs.com there’s quite a stable API (which Fastmail also uses). Patching `prototype` is the only way to update existing functions (as we don’t have access to the original functions). But if someone from Fastmail would be reading
along I’d be happy to find a better way ;-) @brong @rjbs
